// Copyright 2009 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Protocol buffer used by the pagespeed library rule checkers to
// represent violations, potential savings and recommendations.

syntax = "proto2";

package pagespeed;

message Savings {
  // DNS requests that could be saved if page was optimized.
  optional int32 dns_requests_saved = 1;

  // Number of request that could be saved if page was optimized.
  optional int32 requests_saved = 2;

  // Upload bytes that could be saved if page was optimized.
  optional int32 request_bytes_saved = 3;

  // Download bytes that could be saved if page was optimized.
  optional int32 response_bytes_saved = 4;
}

// Base class for result details
// Ids 1000 and above are reserved for custom result details.  All
// custom details must have different ids.
message ResultDetails {
  extensions 1000 to max;
}

message Result {
  // Identifier of rule that produced this result.
  required string rule_name = 1;

  // Potential savings if page was optimized.
  required Savings savings = 3;

  // Resource urls that were processed to compute this result.
  repeated string resource_urls = 4;

  // Full details about the rule violation.
  optional ResultDetails details = 5;
}

message Results {
  repeated Result results = 1;
}
